backend/optimizer.py

```diff
@@ -61,51 +64,75 @@ def basinhopping(self):
+        else:
+            should_keep[-1] = False
 
         return schedule_times, should_keep
 
     def fifty_fifty(self):
         should_keep = [True for x in range(len(self.producer.schedule))]
-        need_grid_power = not self.strictly_positive([s['job'].scheduled_time for s in self.producer.schedule])
+        need_grid_power = not self.strictly_positive()
         if random() < 0.5 or need_grid_power:
             should_keep[-1] = False
         schedule_times = [s['job'].scheduled_time for s in self.producer.schedule]
         return schedule_times, should_keep
 
     def objective_function(self, schedule: List[float]):
+        """The function we want to minimize. schedule is a List of start times for the jobs in the schedule."""
         consumed = defaultdict(float)
-
         for i, load in enumerate(self.differentiated_loads):
             for t, p in load.items():
                 if not math.isnan(schedule[i]):
-                    offset = schedule[i]
-                    consumed[int(round(t + offset))] += p
+                    start_time = schedule[i]
+                    # Add power, p(t), to consumed at time 'start_time' plus 't'
+                    consumed[int(round(start_time + t))] += p
 
-        penalty = 0
-        diff = 0
+        # Differentiate and interpolate production profile on the indices found in consumer
         ts = list(consumed.keys())
         produced = utils.differentiate_and_interpolate(self.producer.prediction, ts)
         ts.extend(list(produced.index.values))
-        ts = sorted(ts)
+        ts = sorted(list(set(ts)))
+
+        total_consumed = 0
+        penalty = 0
         for i, t in enumerate(ts):
-            if i == 0:
-                delta = ts[1] - ts[0]
+            if i == len(ts) - 1:
+                delta = 0
             else:
-                delta = t - ts[i - 1]
+                delta = abs(ts[i+1] - ts[i])
+
             p = produced[t]
             c = consumed[t]
-            diff_t = abs(p - c) * delta
-            diff += diff_t
+
+            # Multiply difference between produced and consumed in time 't' by the distance to the previous time index.
+            # This way, differences are weighted based on their time span.
+            total_consumed += c * delta
+
+            # If consumed is greater than produced, we have negative amount of production power.
             if c > p:
+                diff_t = abs(p - c) * delta
                 penalty += diff_t
 
-        return diff + penalty
+        # Return the difference between produced and consumed energy, in addition to the penalty for having negative
+        # power levels. Finally, subtract this from the objection_value_offset, i.e. the sum of the produced power. This
+        # is so that the theoretically minimal value is equal to zero (if we consume all produced energy).
+        score = abs(self.objection_value_offset - total_consumed) + penalty * self.penalty_factor
```

Comment:
As the comment explains, we now subtract the total power produced. This makes the function equal to zero if we have a perfect solution.